environments/common/src/main/java/org/jboss/weld/environment/deployment/discovery/AbstractDiscoveryStrategy.java

```diff
@@ -47,37 +47,63 @@
  * @author Matej Briškár
  * @author Martin Kouba
  * @author Jozef Hartinger
+ * @author <a href="https://about.me/lairdnelson"
+ * target="_parent">Laird Nelson</a>
  */
 public abstract class AbstractDiscoveryStrategy implements DiscoveryStrategy {
 
-    protected final ResourceLoader resourceLoader;
+    protected ResourceLoader resourceLoader;
 
-    protected final Bootstrap bootstrap;
+    protected Bootstrap bootstrap;
 
-    protected final Set<Class<? extends Annotation>> initialBeanDefiningAnnotations;
+    protected Set<Class<? extends Annotation>> initialBeanDefiningAnnotations;
 
     protected BeanArchiveScanner scanner;
 
     private final List<BeanArchiveHandler> handlers;
 
+    protected AbstractDiscoveryStrategy() {
+        this.handlers = new LinkedList<BeanArchiveHandler>();
+    }
+
     /**
      *
      * @param resourceLoader
      * @param bootstrap
      * @param initialBeanDefiningAnnotations
      */
     public AbstractDiscoveryStrategy(ResourceLoader resourceLoader, Bootstrap bootstrap, Set<Class<? extends Annotation>> initialBeanDefiningAnnotations) {
+        this();
+        this.setResourceLoader(resourceLoader);
+        this.setBootstrap(bootstrap);
+        this.setInitialBeanDefiningAnnotations(initialBeanDefiningAnnotations);
+    }
+
+    @Override
+    public void setResourceLoader(ResourceLoader resourceLoader) {
         this.resourceLoader = resourceLoader;
+    }
+
+    @Override
+    public void setBootstrap(Bootstrap bootstrap) {
         this.bootstrap = bootstrap;
-        this.handlers = new LinkedList<BeanArchiveHandler>();
-        this.initialBeanDefiningAnnotations = initialBeanDefiningAnnotations;
     }
 
     @Override
     public void setScanner(BeanArchiveScanner scanner) {
         this.scanner = scanner;
     }
 
+    public void setInitialBeanDefiningAnnotations(Set<? extends Class<? extends Annotation>> initialBeanDefiningAnnotations) {
+        if (initialBeanDefiningAnnotations == null) {
```

Comment:
I think a simple setter is all we need because it's Weld who provides the set and it's already immutable.